src/util/threadpool.rs

```diff
@@ -169,24 +174,41 @@ impl<Q, T> TaskPool<Q, T>
     }
 }
 
+pub trait Context: Send + Clone {
+    fn set(&mut self, key: &str, value: u64);
+    fn get(&self, key: &str) -> u64;
+    fn on_start(&self);
+    fn on_complete(&self);
+}
+
+pub trait ContextFactory<Ctx: Context> {
+    fn create_context(&self) -> Ctx;
+}
+
+type TTaskPool<Q, T, C> = Arc<(Mutex<TaskPool<Q, T, C>>, Condvar)>;
+
 /// `ThreadPool` is used to execute tasks in parallel.
 /// Each task would be pushed into the pool, and when a thread
 /// is ready to process a task, it will get a task from the pool
 /// according to the `ScheduleQueue` provided in initialization.
-pub struct ThreadPool<Q, T> {
-    task_pool: Arc<(Mutex<TaskPool<Q, T>>, Condvar)>,
+pub struct ThreadPool<Q, T, C, Ctx> {
+    task_pool: TTaskPool<Q, T, Ctx>,
     threads: Vec<JoinHandle<()>>,
     task_count: Arc<AtomicUsize>,
-    sender: Sender<Task<T>>,
+    sender: Sender<Task<T, Ctx>>,
+    // ctx_factory should only be used in one thread
```

Comment:
In current implementation of coprocessor, only one thread invoke threadpool.execute, which will use ctx_factory.